logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java

```diff
@@ -212,10 +266,92 @@ private void innerWriteEntry(DLQEntry entry) throws IOException {
         if (currentWriter.getPosition() + eventPayloadSize > maxSegmentSize) {
```

Comment:
What happens in the `do...while` loop if a segment is cleaned? 

Can we end in a situation where we either never exit the loop, we crash logstash, or we delete more than expected?

Reply:
Actually we can crash, we could go in infinite loop only if we don't update the size of the DLQ.
We can delete more data than expected by just applying the age policy, but it deletes more data because the maxQueueSize` constraint is not respected. So it could be that the two limitation mechanism act at the same time.
Before kicks in the age retention policy to drop expired messages, then the max size constraint for FIFO queue type could drop even more segments.